### PR TITLE
Issue/5349 cleanup old agents

### LINKED ISSUE #5349: Cleanup old agents from agent table in db
The agent table in the database is never cleaned up properly. It contains agents that were present in a very old version of the configuration model, while the currently released version doesn't have any resources for it anymore. This ticket should make sure that the agent table is properly cleaned up.

An agent can be deleted iff:
- it's not used by any version
- it's not in the agent map
- it's down

Cleanup should be done right after resource cleanup.

### PR DESCRIPTION
**Pull request by FloLey**
# Description

The agent table in the database was never cleaned up properly. It contained agents that were present in a very old version of the configuration model, while the currently released version doesn't have any resources for it anymore. This PR makes surre that old agents are cleaned up.

An agent can be deleted iff:

1. it's not used by any version
2. it's not in the agent map
3. it's down

closes #5349

# Self Check:

Strike through any lines that are not applicable (`~~line~~`) then check the box

- [x] Attached issue to pull request
- [x] Changelog entry
- [x] Type annotations are present
- [x] Code is clear and sufficiently documented
- [x] No (preventable) type errors (check using make mypy or make mypy-diff)
- [x] Sufficient test cases (reproduces the bug/tests the requested feature)
- [x] Correct, in line with design
- [ ] End user documentation is included or an issue is created for end-user documentation (add ref to issue here: )

### CODE REVIEW

File: tests/test_server.py [hunk 1]
```diff
@@ -1405,3 +1405,95 @@ async def test_redirect_dashboard_to_console(server, path):
...
+    agents_before_purge = await data.Agent.get_list()
+    assert len(agents_before_purge) == 6
+
+    await server.get_slice(SLICE_ORCHESTRATION)._purge_versions()
+    agents_after_purge = await data.Agent.get_list()
+    assert len(agents_after_purge) == 4
+    assert any(agent.name == "agent2" and agent.environment == env1.id for agent in agents_after_purge)
+    assert any(agent.name == "agent3" and agent.environment == env1.id for agent in agents_after_purge)
+    assert any(agent.name == "agent4" and agent.environment == env1.id for agent in agents_after_purge)
+    assert any(agent.name == "agent1" and agent.environment == env2.id for agent in agents_after_purge)
```
Comment (arnaudsjs) [inline tests/test_server.py:1499]: These checks are very verbose and also very hard to read:
```suggestion
    agents_after_purge = [(agent.environment, agent.name) for agent in await data.Agent.get_list()]
    assert len(agents_after_purge) == 4
    expected_agents_after_purge = [
        (env1.id, "agent2"),
        (env1.id, "agent3"),
        (env1.id, "agent4"),
        (env2.id, "agent1"),
    ]
    assert sorted(agents_after_purge) == sorted(expected_agents_after_purge)
```

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/server/services/orchestrationservice.py [hunk 2]
```diff
@@ -384,6 +384,39 @@ async def _purge_versions(self) -> None:
...
+        FROM agent
+        WHERE id_primary IS not NULL
+
+        -- Union with the set of (environment, name) pairs for all resources
+        -- and associated agents for each environment. (that are used by a version)
+        UNION ALL
+        SELECT r.environment, r.agent
+        FROM public.resource r
+        JOIN public.configurationmodel c ON r.environment = c.environment AND r.model = c.version
+        GROUP BY r.environment, r.agent
```
[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/server/services/orchestrationservice.py]: but it is public.resource that owns a FK that references a configuration model right?

File: src/inmanta/server/services/orchestrationservice.py [hunk 3]
```diff
@@ -384,6 +384,39 @@ async def _purge_versions(self) -> None:
...
+    SELECT DISTINCT environment_id as environment, agent as name
+    FROM (
+        -- agent is in the agent map
+        SELECT e.id as environment_id, map.key as agent
+        FROM environment e
+        CROSS JOIN LATERAL jsonb_each(e.settings->'autostart_agent_map') AS map(key, value)
+
+        -- Union with the set of (environment, name) pairs for all agents that
+        -- have a primary ID set (that are not down)
+        UNION ALL
```
[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/server/services/orchestrationservice.py:404]: I thought that was how you could join the results of multiple queries but it seems, just UNION is better here as it will remove the duplicates already

File: src/inmanta/server/services/orchestrationservice.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/services/orchestrationservice.py]: Yes. You are right. I said it incorrectly. I mean that if you have a one to many relationship between two tables, the query becomes more readable when you start from the table on the side with arity 1.

File: src/inmanta/server/services/orchestrationservice.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/services/orchestrationservice.py:404]: I agree.

File: src/inmanta/server/services/orchestrationservice.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/server/services/orchestrationservice.py]: I see and I think I can see why. I'll try to pay attention to it

File: tests/test_server.py [hunk 4]
```diff
@@ -1405,3 +1405,95 @@ async def test_redirect_dashboard_to_console(server, path):
     )
     response = await http_client.fetch(request, raise_error=False)
     assert result_url == response.effective_url
+
+
+async def test_cleanup_old_agents(server):
+    """
+    This test is testing the functionality of cleaning up old agents in the database that are no longer present in the
+    currently released version of the configuration model. The test creates 2 environments and adds agents with various
```
[the remark this answers is not in the document]
Comment (FloLey) [reply tests/test_server.py:1413]: Is this comment wrong or do you find it unnecessary/unclear?

[the remark this answers is not in the document]
Comment (arnaudsjs) [reply tests/test_server.py:1413]: The old sentence seems to indicate that the cleanup conditions is that the agent is no longer used in the currently released version of the model. But that is not correct.

File: src/inmanta/server/services/orchestrationservice.py [hunk 5]
```diff
@@ -384,6 +384,39 @@ async def _purge_versions(self) -> None:
                 for v in delete_list:
                     await version_dict[v].delete_cascade()
 
+        # Cleanup old agents from agent table in db
+        async with data.Agent.get_connection() as con:
+            await con.execute(
+                """
+-- Delete all records from the agent table where the (environment, name) pair
+-- is not in the set of (environment, name) pairs returned by the subquery.
+DELETE FROM public.agent
```
[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/server/services/orchestrationservice.py:393]: I did not. Could you repeat to me how I can check this? I'll add it to the dev docs so I won't forget it next time

File: changelogs/unreleased/5349-cleanup-old-agents.yml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+---
+description: cleanup old agents from the agent table in db
```
Comment (Hugo-Inmanta) [inline changelogs/unreleased/5349-cleanup-old-agents.yml:2]: I would add a capital letter and a dot so that it looks better in the final changelog, and maybe rephrase it slightly: (not a strong opinion) 
```suggestion
description: Old agents in the agent table are now cleaned up from the database.
```

**Review by Hugo-Inmanta — APPROVED** [review]
This looks ok as far as I can tell, but then again query crafting is not my forte

File: src/inmanta/server/services/orchestrationservice.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/services/orchestrationservice.py]: You can restore a database dump from the tests folder on your local PostgreSQL installation and run the query against that database but with the EXPLAIN ANALYZE prefix. That way postgresql will report which indexes it uses and what the estimated cost of each query is.

[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/server/services/orchestrationservice.py]: thanks! this is the result:
```
  Delete on agent  (cost=20.00..21.03 rows=1 width=6) (actual time=0.066..0.068 rows=0 loops=1)
   ->  Seq Scan on agent  (cost=20.00..21.03 rows=1 width=6) (actual time=0.065..0.067 rows=0 loops=1)
         Filter: (NOT (hashed SubPlan 1))
         Rows Removed by Filter: 2
         SubPlan 1
           ->  HashAggregate  (cost=17.50..19.50 rows=200 width=48) (actual time=0.044..0.045 rows=2 loops=1)
                 Group Key: e.id, map.key
                 Batches: 1  Memory Usage: 40kB
                 ->  HashAggregate  (cost=12.35..14.41 rows=206 width=48) (actual time=0.042..0.044 rows=2 loops=1)
                       Group Key: e.id, map.key
                       Batches: 1  Memory Usage: 40kB
                       ->  Append  (cost=0.01..11.32 rows=206 width=48) (actual time=0.012..0.037 rows=6 loops=1)
                             ->  Nested Loop  (cost=0.01..5.02 rows=200 width=48) (actual time=0.012..0.013 rows=2 loops=1)
                                   ->  Seq Scan on environment e  (cost=0.00..1.02 rows=2 width=48) (actual time=0.002..0.002 rows=2 loops=1)
                                   ->  Function Scan on jsonb_each map  (cost=0.01..1.00 rows=100 width=32) (actual time=0.004..0.004 rows=1 loops=2)
                             ->  Subquery Scan on "*SELECT* 2"  (cost=0.00..1.04 rows=2 width=48) (actual time=0.001..0.002 rows=2 loops=1)
                                   ->  Seq Scan on agent agent_1  (cost=0.00..1.02 rows=2 width=48) (actual time=0.001..0.001 rows=2 loops=1)
                                         Filter: (id_primary IS NOT NULL)
                             ->  Subquery Scan on "*SELECT* 3"  (cost=2.15..2.22 rows=4 width=48) (actual time=0.019..0.021 rows=2 loops=1)
                                   ->  Unique  (cost=2.15..2.18 rows=4 width=48) (actual time=0.019..0.020 rows=2 loops=1)
                                         ->  Sort  (cost=2.15..2.16 rows=4 width=48) (actual time=0.019..0.020 rows=4 loops=1)
                                               Sort Key: r.environment, r.agent
                                               Sort Method: quicksort  Memory: 25kB
                                               ->  Hash Join  (cost=1.05..2.11 rows=4 width=48) (actual time=0.009..0.010 rows=4 loops=1)
                                                     Hash Cond: ((r.environment = c.environment) AND (r.model = c.version))
                                                     ->  Seq Scan on resource r  (cost=0.00..1.04 rows=4 width=52) (actual time=0.001..0.001 rows=4 loops=1)
                                                     ->  Hash  (cost=1.02..1.02 rows=2 width=20) (actual time=0.004..0.004 rows=2 loops=1)
                                                           Buckets: 1024  Batches: 1  Memory Usage: 9kB
                                                           ->  Seq Scan on configurationmodel c  (cost=0.00..1.02 rows=2 width=20) (actual time=0.002..0.002 rows=2 loops=1)
 Planning Time: 0.261 ms
 Execution Time: 0.133 ms
(31 rows)
```
So indeed the third select seems to cost more and the NOT has some cost to, but do I understand correctly that the Union call is the one that cost the most? Also I'm not able to tell how "bad" it is

[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/services/orchestrationservice.py]: Indeed. It's hard to tell from this how bad it is. Especially because the data set is rather small. I did the same on a dump of the database from mgmt and then we get the following:

```
 Seq Scan on agent  (cost=6618.02..6623.90 rows=96 width=69) (actual time=61.036..61.064 rows=46 loops=1)
   Filter: (NOT (hashed SubPlan 1))
   Rows Removed by Filter: 146
   SubPlan 1
     ->  Unique  (cost=6610.84..6617.52 rows=200 width=48) (actual time=60.940..60.973 rows=193 loops=1)
           ->  Sort  (cost=6610.84..6613.07 rows=890 width=48) (actual time=60.939..60.950 rows=380 loops=1)
                 Sort Key: e.id, map.key
                 Sort Method: quicksort  Memory: 58kB
                 ->  Append  (cost=0.01..6558.34 rows=890 width=48) (actual time=0.048..60.651 rows=380 loops=1)
                       ->  Nested Loop  (cost=0.01..13.06 rows=600 width=48) (actual time=0.047..0.194 rows=186 loops=1)
                             ->  Seq Scan on environment e  (cost=0.00..1.06 rows=6 width=48) (actual time=0.008..0.009 rows=6 loops=1)
                             ->  Function Scan on jsonb_each map  (cost=0.01..1.00 rows=100 width=32) (actual time=0.026..0.027 rows=31 loops=6)
                       ->  Subquery Scan on "*SELECT* 2"  (cost=0.00..6.35 rows=143 width=48) (actual time=0.003..0.022 rows=143 loops=1)
                             ->  Seq Scan on agent agent_1  (cost=0.00..4.92 rows=143 width=43) (actual time=0.002..0.014 rows=143 loops=1)
                                   Filter: (id_primary IS NOT NULL)
                                   Rows Removed by Filter: 49
                       ->  Subquery Scan on "*SELECT* 3"  (cost=6525.54..6528.48 rows=147 width=48) (actual time=60.413..60.419 rows=51 loops=1)
                             ->  HashAggregate  (cost=6525.54..6527.01 rows=147 width=44) (actual time=60.412..60.416 rows=51 loops=1)
                                   Group Key: r.environment, r.agent
                                   Batches: 1  Memory Usage: 40kB
                                   ->  Nested Loop  (cost=0.42..4722.16 rows=360675 width=44) (actual time=0.019..28.297 rows=360622 loops=1)
                                         ->  Seq Scan on configurationmodel c  (cost=0.00..13.73 rows=173 width=20) (actual time=0.001..0.028 rows=173 loops=1)
                                         ->  Index Only Scan using resource_env_model_agent_index on resource r  (cost=0.42..19.36 rows=786 width=48) (actual time=0.002..0.087 rows=2085 loops=173)
                                               Index Cond: ((environment = c.environment) AND (model = c.version))
                                               Heap Fetches: 0
 Planning Time: 2.399 ms
 Execution Time: 61.304 ms
(27 rows)
```
Then it becomes very clear that the last sub-query is very expensive. Also note that every single sub-query is doing a sequential table scan. We really have to improve the index usage. A few ideas to make it perform better:

1) Move the `is_primary` check outside of the NOT IN statement. That way we can trim down the amount of potential agents quickly without performing the expensive NOT IN check.

```
DELETE FROM public.agent
id_primary IS NULL AND ...
```

2) The sub-query that checks whether an agent is still referenced by a resource, performs a join between configurationmodel and resource. I don't think that join is required. It's the resource that requires the agent and not the model. So we could do something like this:

```
DELETE FROM public.agent AS a
WHERE NOT EXISTS(SELECT 1 FROM resource AS re where a.environment=re.environment AND a.name=re.agent)
```
If we do that, we need to make sure that the resource table has an index on (environment, agent).

[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/services/orchestrationservice.py]: I am going to create a ticket to make it easier to do performance tests of queries, because the current approach is not very developer friendly.

[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/services/orchestrationservice.py]: I think there is a bug in these sub-queries. They don't filter on the environment that is used by the surrounding Agent record. Is that correct?

[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/server/services/orchestrationservice.py]: I'm not sure I understand the bug you are referring to, but I've updated the query so maybe I've solved it?

[the remark this answers is not in the document]
Comment (FloLey) [reply src/inmanta/server/services/orchestrationservice.py]: @arnaudsjs regarding the index there is already this one:
`CREATE INDEX resource_env_model_agent_index ON resource (environment, model, agent);`
Do you think it is worth it to still add another one without the model?

[the remark this answers is not in the document]
Comment (arnaudsjs) [reply src/inmanta/server/services/orchestrationservice.py]: An index on `(environment, model, agent)` cannot be used to do a lookup for a `(environment, agent)` pair, is it? You can only use that index to filter on `environment`, but that column is already part of the primary key.

File: src/inmanta/data/__init__.py [hunk 7]
```diff
@@ -3358,6 +3358,31 @@ async def mark_all_as_non_primary(cls, connection: Optional[asyncpg.connection.C
...
+        query = """
+DELETE FROM public.agent AS a
+WHERE (environment, name) NOT IN (
+    SELECT DISTINCT environment_id as environment, agent as name
+    FROM (
+        -- agent is in the agent map
+        SELECT e.id as environment_id, map.key as agent
+        FROM public.environment e
+        CROSS JOIN LATERAL jsonb_each(e.settings->'autostart_agent_map') AS map(key, value)
+    ) in_agent_map
```
Comment (arnaudsjs) [inline src/inmanta/data/__init__.py:3372]: This sub-query must filter on the correct environment.

File: changelogs/unreleased/5349-cleanup-old-agents.yml [hunk 8]
```diff
@@ -0,0 +1,7 @@
+---
+description: Old agents in the agent table are now cleaned up from the database.
+issue-nr: 5349
+change-type: patch
+destination-branches: [master, iso6, iso5]
```
Comment (arnaudsjs) [inline changelogs/unreleased/5349-cleanup-old-agents.yml:5]: This should not go into iso5

**Review by arnaudsjs — CHANGES_REQUESTED** [review]

File: src/inmanta/data/__init__.py (hunk 7 above)
Comment (arnaudsjs) [reply src/inmanta/data/__init__.py:3372]: I seems I was too quick here. This query seems to perform better than my suggestion. You can ignore this one.

Comment (arnaudsjs) [reply src/inmanta/data/__init__.py:3372]: Ah. I see. You are using the (environment, name) index on the agent table here.

**Review by arnaudsjs — APPROVED** [review]
